### PR TITLE
[InstSimplify] Fold icmp of `X and/or C1` and `X and/or C2` into constant

### LINKED ISSUE #65833: Simplifications of `max/min (a & c0, a & c1)` under `(c0 & c1) == c1/c0`
The following two functions can be simplified, and this is motivated from https://gcc.gnu.org/bugzilla/show_bug.cgi?id=109878.

```cpp
#include<iostream>

unsigned foo(unsigned a, unsigned c0, unsigned c1)
{
   return ((c0 & c1) == c1) ? std::max(a & c0, a & c1) : (a & c0);
}

unsigned bar(unsigned a, unsigned c0, unsigned c1)
{
   return ((c0 & c1) == c0) ? std::min(a & c0, a & c1) : (a & c0);
}
```

Alive2:
https://alive2.llvm.org/ce/z/kQrReA
https://alive2.llvm.org/ce/z/UQay5M

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch simplifies the pattern `icmp X and/or C1, X and/or C2` when one constant mask is the subset of the other.
If `C1 & C2 == C1`, `A = X and/or C1`, `B = X and/or C2`, we can do the following folds:
`icmp ule A, B -> true`
`icmp ugt A, B -> false`
We can apply similar folds for signed predicates when `C1` and `C2` are the same sign:
`icmp sle A, B -> true`
`icmp sgt A, B -> false`

Alive2: https://alive2.llvm.org/ce/z/Q4ekP5
Fixes #65833.

### CODE REVIEW

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 1]
```diff
@@ -3436,6 +3436,32 @@ static Value *simplifyICmpWithBinOp(CmpInst::Predicate Pred, Value *LHS,
       if (Value *V = simplifyICmpInst(Pred, LBO->getOperand(1),
                                       RBO->getOperand(1), Q, MaxRecurse - 1))
         return V;
+      break;
+    }
+    // icmp X & C1, X & C2 where (C1 & C2) == C1/C2
+    // icmp X | C1, X | C2 where (C1 & C2) == C1/C2
```
Comment (goldsteinn) [inline llvm/lib/Analysis/InstructionSimplify.cpp:3442]: In the title and here please fill in what you are transforming to.
Title also should indicate it works for `or` as well.

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 2]
```diff
@@ -3436,6 +3436,32 @@ static Value *simplifyICmpWithBinOp(CmpInst::Predicate Pred, Value *LHS,
...
+    case Instruction::And:
+    case Instruction::Or: {
+      if (ICmpInst::isUnsigned(Pred)) {
+        const APInt *C1, *C2;
+        if (match(LBO->getOperand(1), m_APInt(C1)) &&
+            match(RBO->getOperand(1), m_APInt(C2))) {
+          if (C1->isSubsetOf(*C2)) {
```
Comment (goldsteinn) [inline llvm/lib/Analysis/InstructionSimplify.cpp:3449]: Think cleaner would be something like:

```
if(!C1->isSubsetOf(*C2)) {
    swap(C1, C2);
    Pred = getSwappedPred(Pred);
}
if(C1->isSubsetOf(*C2)) {
....
}
```

So you don't need to duplicate logic.

File: llvm/test/Transforms/InstSimplify/maxmin_intrinsics.ll [hunk 3]
```diff
@@ -2332,3 +2332,100 @@ false:
...
+; CHECK-NEXT:    [[AND1:%.*]] = or i8 [[X:%.*]], 1
+; CHECK-NEXT:    [[AND2:%.*]] = or i8 [[X]], 2
+; CHECK-NEXT:    [[VAL:%.*]] = call i8 @llvm.umax.i8(i8 [[AND1]], i8 [[AND2]])
+; CHECK-NEXT:    ret i8 [[VAL]]
+;
+  %and1 = or i8 %x, 1
+  %and2 = or i8 %x, 2
+  %val = call i8 @llvm.umax.i8(i8 %and1, i8 %and2)
+  ret i8 %val
+}
```
Comment (goldsteinn) [inline llvm/test/Transforms/InstSimplify/maxmin_intrinsics.ll:2431]: You know the drill for tests...

**Comment by goldsteinn** [comment]
LGTM, but please cleanup commits.

**Review by goldsteinn — APPROVED** [review]

**Comment by dtcxzyw** [comment]
> LGTM, but please cleanup commits.

GitHub will squash commits into a single commit.

**Pull request merged**